File: lib/ontology.ts
/**
 * 온톨로지 코어 (app/ontology.py → TypeScript)
 * Gold 데이터를 객체·관계로 승격, YAML 스키마 기반 액션 스코어링
 */
import type { SupabaseClient } from '@supabase/supabase-js'
import { load as yamlLoad } from 'js-yaml'
import { readFileSync } from 'fs'
import { join } from 'path'

interface OntologySchema {
  facility_filter: string
  actions: Record<string, {
    name: string
    description: string
    weights: { youth_pop: number; employees: number; facility_gap: number; outmigration: number }
  }>
  keyword_mapping: Record<string, string[]>
}

let _schema: OntologySchema | null = null

function getSchema(): OntologySchema {
  if (!_schema) {
    const p = join(process.cwd(), 'config', 'ontology-schema.yaml')
    _schema = yamlLoad(readFileSync(p, 'utf8')) as OntologySchema
  }
  return _schema
}

export async function buildOntology(supabase: SupabaseClient) {
  const schema = getSchema()

  // 기존 객체/관계 초기화
  await supabase.from('onto_objects').delete().neq('obj_id', 'NEVER_MATCH')
  await supabase.from('onto_links').delete().neq('src', 'NEVER_MATCH')

  const { data: tenants } = await supabase.from('tenants').select('sgg_cd,name,gov_type')
  const objs: Array<{ obj_id: string; obj_type: string; label: string; props: string }> = []
  const links: Array<{ src: string; rel: string; dst: string; weight: number }> = []

  for (const t of tenants ?? []) {
    objs.push({ obj_id: `sigun:${t.sgg_cd}`, obj_type: '시군', label: t.name, props: `유형=${t.gov_type}` })
  }

  const { data: yrData } = await supabase
    .from('gold_youth_population').select('year').order('year', { ascending: false }).limit(1)
  const latest = yrData?.[0]?.year
  if (!latest) return { objects: 0, links: 0, year: null }

  const { data: yp } = await supabase.rpc('agg_youth_pop', { yr: latest })
  for (const r of yp ?? []) {
    const net = Number(r.inf) - Number(r.outf)
    const oid = `youth:${r.sgg_cd}`
    objs.push({ obj_id: oid, obj_type: '청년인구', label: `${r.sigun} 청년`, props: `인구=${r.pop};순이동=${net}` })
    links.push({ src: `sigun:${r.sgg_cd}`, rel: '청년규모', dst: oid, weight: Number(r.pop) })
    links.push({ src: oid, rel: net >= 0 ? '순유입' : '순유출', dst: `sigun:${r.sgg_cd}`, weight: Math.abs(net) })
  }

  const { data: bz } = await supabase.rpc('agg_business', { yr: latest })
  for (const r of bz ?? []) {
    const oid = `biz:${r.sgg_cd}`
    objs.push({ obj_id: oid, obj_type: '사업체', label: `${r.sigun} 사업체`, props: `사업체=${r.bc};종사자=${r.emp}` })
    links.push({ src: `sigun:${r.sgg_cd}`, rel: '산업기반', dst: oid, weight: Number(r.emp) })
  }

  const { data: fac } = await supabase.rpc('agg_facility', { ftype_filter: schema.facility_filter })
  for (const r of fac ?? []) {
    const oid = `fac:${r.sgg_cd}`
    objs.push({ obj_id: oid, obj_type: '청년인프라', label: `${r.sigun} ${schema.facility_filter}`, props: `개수=${r.n}` })
    links.push({ src: `sigun:${r.sgg_cd}`, rel: '보유시설', dst: oid, weight: Number(r.n) })
  }

  if (objs.length)  await supabase.from('onto_objects').insert(objs)
  if (links.length) await supabase.from('onto_links').insert(links)
  return { objects: objs.length, links: links.length, year: latest }
}

export async function getGraph(supabase: SupabaseClient, centerSgg?: string) {
  if (centerSgg) {
    const oids = [`sigun:${centerSgg}`, `youth:${centerSgg}`, `biz:${centerSgg}`, `fac:${centerSgg}`]
    const { data: nodes } = await supabase.from('onto_objects').select('*').in('obj_id', oids)
    const { data: edges } = await supabase.from('onto_links').select('*')
      .or(`src.in.(${oids.join(',')}),dst.in.(${oids.join(',')})`)
    return { nodes: nodes ?? [], edges: edges ?? [] }
  }
  const { data: nodes } = await supabase.from('onto_objects').select('*')
  const { data: edges } = await supabase.from('onto_links').select('*')
  return { nodes: nodes ?? [], edges: edges ?? [] }
}

export function recommendOntologyCandidates(meta: Record<string, unknown>) {
  const schema = getSchema()
  const text = [meta.title, meta.description, meta.theme, meta.keywords]
    .map(v => String(v ?? '')).join(' ').toLowerCase()

  const results: Array<{ obj_type: string; matched_keywords: string[]; match_score: number; reason: string }> = []
  for (const [objType, kws] of Object.entries(schema.keyword_mapping)) {
    const matched = kws.filter(kw => text.includes(String(kw)))
    if (matched.length) {
      results.push({
        obj_type: objType,
        matched_keywords: matched,
        match_score: matched.length,
        reason: `'${matched.slice(0, 3).join(', ')}' 키워드가 메타데이터에서 발견됨`,
      })
    }
  }
  return results.sort((a, b) => b.match_score - a.match_score)
}

export function listActions() {
  const schema = getSchema()
  return Object.entries(schema.actions).map(([key, v]) => ({
    key, name: v.name, description: v.description,
  }))
}

export async function scoreAction(supabase: SupabaseClient, actionKey: string, top = 10) {
  const schema = getSchema()
  const action = schema.actions[actionKey]
  if (!action) return []

  const { data: yrData } = await supabase
    .from('gold_youth_population').select('year').order('year', { ascending: false }).limit(1)
  const latest = yrData?.[0]?.year
  if (!latest) return []

  const { data: rows } = await supabase.rpc('score_action_data', {
    yr: latest, ftype_filter: schema.facility_filter,
  })
  if (!rows?.length) return []

  const w = action.weights
  const mxPop = Math.max(...rows.map((r: any) => Number(r.pop))) || 1
  const mxEmp = Math.max(...rows.map((r: any) => Number(r.emp))) || 1
  const mxFac = Math.max(...rows.map((r: any) => Number(r.fac))) || 1
  const mxOut = Math.max(...rows.map((r: any) => Math.abs(Math.min(0, Number(r.net))))) || 1

  const out = rows.map((r: any) => {
    const pop = Number(r.pop); const net = Number(r.net)
    const emp = Number(r.emp); const fac = Number(r.fac)
    const score = (
      w.youth_pop    * (pop / mxPop) +
      w.employees    * (emp / mxEmp) +
      w.facility_gap * (1 - fac / mxFac) +
      w.outmigration * (Math.abs(Math.min(0, net)) / mxOut)
    ) * 100
    return {
      sgg_cd: r.sgg_cd, sigun: r.sigun,
      youth_pop: pop, net_migration: net, employees: emp, youth_centers: fac,
      priority_score: Math.round(score * 10) / 10,
    }
  }).sort((a: any, b: any) => b.priority_score - a.priority_score)

  return out.slice(0, top).map((o: any, i: number) => ({ ...o, rank: i + 1 }))
}

File: app/ontology.py
"""
이음(EUM) 플랫폼 - 온톨로지 코어 (L4, 2단계 코어의 MVP)
Gold 데이터를 객체(Object)·관계(Link)로 승격한다.
객체: 시군, 청년인구(시군x연도 집계), 사업체(시군 집계), 공공시설
관계: 시군-(보유)->시설, 시군-(청년규모)->청년인구, 청년인구-(유입/유출)
액션: 정착지원 우선지역 선정 (스코어링)
"""
from . import database as db


def build_ontology():
    """Gold 데이터에서 객체/관계를 생성(멱등)."""
    db.execute("DELETE FROM onto_objects")
    db.execute("DELETE FROM onto_links")

    objs = []
    links = []

    # 1) 시군 객체
    for t in db.query("SELECT sgg_cd, name, gov_type FROM tenants"):
        oid = f"sigun:{t['sgg_cd']}"
        objs.append((oid, "시군", t["name"], f"유형={t['gov_type']}"))

    # 2) 청년인구 객체 (시군 x 최신연도 집계) + 관계
    latest = db.query("SELECT max(year) y FROM gold_youth_population")[0]["y"]
    yp = db.query("""
        SELECT sgg_cd, sigun, sum(population) pop, sum(inflow) inf, sum(outflow) outf
        FROM gold_youth_population WHERE year = ?
        GROUP BY sgg_cd, sigun""", [latest])
    for r in yp:
        oid = f"youth:{r['sgg_cd']}"
        net = r["inf"] - r["outf"]
        objs.append((oid, "청년인구", f"{r['sigun']} 청년",
                     f"인구={r['pop']};순이동={net}"))
        links.append((f"sigun:{r['sgg_cd']}", "청년규모", oid, float(r["pop"])))
        links.append((oid, "순유입" if net >= 0 else "순유출",
                      f"sigun:{r['sgg_cd']}", float(abs(net))))

    # 3) 사업체 객체(시군 집계) + 관계
    bz = db.query("""
        SELECT sgg_cd, sigun, sum(biz_count) bc, sum(employees) emp
        FROM gold_business WHERE year = ? GROUP BY sgg_cd, sigun""", [latest])
    for r in bz:
        oid = f"biz:{r['sgg_cd']}"
        objs.append((oid, "사업체", f"{r['sigun']} 사업체",
                     f"사업체={r['bc']};종사자={r['emp']}"))
        links.append((f"sigun:{r['sgg_cd']}", "산업기반", oid, float(r["emp"])))

    # 4) 공공시설 객체(청년센터만 대표로) + 관계
    fac = db.query("""
        SELECT sgg_cd, sigun, count(*) n
        FROM gold_public_facility WHERE ftype = '청년센터'
        GROUP BY sgg_cd, sigun""")
    for r in fac:
        oid = f"fac:{r['sgg_cd']}"
        objs.append((oid, "청년인프라", f"{r['sigun']} 청년센터",
                     f"개수={r['n']}"))
        links.append((f"sigun:{r['sgg_cd']}", "보유시설", oid, float(r["n"])))

    for o in objs:
        db.execute("INSERT INTO onto_objects VALUES (?,?,?,?)", list(o))
    for l in links:
        db.execute("INSERT INTO onto_links VALUES (?,?,?,?)", list(l))

    return {"objects": len(objs), "links": len(links), "year": latest}


def graph(center_sgg=None):
    """그래프 시각화용 노드/엣지 반환. center_sgg가 있으면 해당 시군 중심."""
    if center_sgg:
        oids = [f"sigun:{center_sgg}", f"youth:{center_sgg}",
                f"biz:{center_sgg}", f"fac:{center_sgg}"]
        ph = ",".join(["?"] * len(oids))
        nodes = db.query(f"SELECT * FROM onto_objects WHERE obj_id IN ({ph})", oids)
        edges = db.query(
            f"SELECT * FROM onto_links WHERE src IN ({ph}) OR dst IN ({ph})",
            oids + oids)
    else:
        nodes = db.query("SELECT * FROM onto_objects")
        edges = db.query("SELECT * FROM onto_links")
    return {"nodes": nodes, "edges": edges}


# 온톨로지 객체 유형별 매칭 키워드
_OBJ_TYPE_KEYWORDS = {
    "청년인구": ["청년", "인구", "유입", "유출", "이동", "유동인구", "생활인구", "청년인구"],
    "사업체":   ["사업체", "기업", "업체", "종사자", "고용", "산업", "제조업", "사업장"],
    "청년인프라": ["시설", "센터", "인프라", "청년시설", "공공시설", "청년센터", "복지관"],
    "시군":     ["시군", "시군구", "지역", "행정구역", "거주", "주소", "행정동", "읍면동"],
}


def recommend_ontology_candidates(meta: dict) -> list[dict]:
    """제출 메타데이터를 키워드 분석해 온톨로지 객체 유형 후보를 추천한다.
    meta: submissions 테이블 행 dict — title/description/theme/keywords 사용."""
    text = " ".join([
        str(meta.get("title") or ""),
        str(meta.get("description") or ""),
        str(meta.get("theme") or ""),
        str(meta.get("keywords") or ""),
    ]).lower()

    results = []
    for obj_type, kws in _OBJ_TYPE_KEYWORDS.items():
        matched = [kw for kw in kws if kw in text]
        if matched:
            results.append({
                "obj_type": obj_type,
                "matched_keywords": matched,
                "reason": f"'{', '.join(matched[:3])}' 키워드가 메타데이터에서 발견됨",
            })

    return results


def action_settlement_priority(top=10):
    """
    액션: 청년 정착지원 우선지역 선정.
    온톨로지 객체 속성으로 스코어링(데모 휴리스틱).
    점수 = 청년규모 정규화 + 산업기반 + 청년인프라 - 순유출 패널티
    """
    latest = db.query("SELECT max(year) y FROM gold_youth_population")[0]["y"]
    rows = db.query("""
        WITH y AS (
            SELECT sgg_cd, sigun, sum(population) pop,
                   sum(inflow)-sum(outflow) net
            FROM gold_youth_population WHERE year=? GROUP BY sgg_cd, sigun),
        b AS (
            SELECT sgg_cd, sum(employees) emp FROM gold_business
            WHERE year=? GROUP BY sgg_cd),
        f AS (
            SELECT sgg_cd, count(*) fac FROM gold_public_facility
            WHERE ftype='청년센터' GROUP BY sgg_cd)
        SELECT y.sgg_cd, y.sigun, y.pop, y.net,
               coalesce(b.emp,0) emp, coalesce(f.fac,0) fac
        FROM y LEFT JOIN b ON y.sgg_cd=b.sgg_cd
               LEFT JOIN f ON y.sgg_cd=f.sgg_cd
    """, [latest, latest])
    if not rows:
        return []
    mx_pop = max(r["pop"] for r in rows) or 1
    mx_emp = max(r["emp"] for r in rows) or 1
    mx_fac = max(r["fac"] for r in rows) or 1
    mx_out = max(abs(min(0, r["net"])) for r in rows) or 1
    out = []
    for r in rows:
        pop_s = r["pop"] / mx_pop
        emp_s = r["emp"] / mx_emp
        fac_gap = 1 - (r["fac"] / mx_fac)         # 인프라 부족할수록 우선
        out_pen = abs(min(0, r["net"])) / mx_out  # 순유출 클수록 우선
        score = (0.30 * pop_s + 0.20 * emp_s +
                 0.25 * fac_gap + 0.25 * out_pen) * 100
        out.append({
            "sgg_cd": r["sgg_cd"], "sigun": r["sigun"],
            "youth_pop": r["pop"], "net_migration": r["net"],
            "employees": r["emp"], "youth_centers": r["fac"],
            "priority_score": round(score, 1),
        })
    out.sort(key=lambda x: x["priority_score"], reverse=True)
    for i, o in enumerate(out, 1):
        o["rank"] = i
    return out[:top]
